specfile/prep.py

```diff
@@ -0,0 +1,303 @@
+                    result.append(macro)
+                    # add the second macro
+                    macro = PatchMacro(i, "%patch")
+                    macro._parse_options(options)
+                    result.append(macro)
+                else:
+                    result.append(macro)
+            elif name == "%autosetup":
+                macro = AutosetupMacro(i, name)
+                macro._parse_options(options)
+                result.append(macro)
+            elif name == "%autopatch":
+                macro = AutopatchMacro(i, name)
+                macro._parse_options(options)
+                result.append(macro)
+        return result
+
+    @overload
+    def __getitem__(self, i: int) -> PrepMacro:
+        pass
+
+    @overload
+    def __getitem__(self, i: slice) -> List[PrepMacro]:
+        pass
+
+    def __getitem__(self, i):
+        return self._get_items()[i]
+
+
+class Prep:
+    """
+    Class that represents a %prep section.
+
+    Attributes:
+        macros: Sequence of individual %prep macros.
+            Recognized macros are %setup, %patch, %autosetup and %autopatch.
+    """
+
+    def __init__(self, section: Section) -> None:
+        """
+        Constructs a `Prep` object.
+
+        Args:
+            section: %prep section.
+
+        Returns:
+            Constructed instance of `Prep` class.
+        """
+        self._section = section
+        self.macros = PrepMacros(self._section)
+
+    def __repr__(self) -> str:
+        section = repr(self._section)
+        return f"Prep({section})"
+
+    def add_patch_macro(
+        self,
+        index: int,
+        P: Optional[int] = None,
+        p: Optional[int] = None,
+        b: Optional[str] = None,
+        E: Optional[bool] = None,
+    ) -> None:
+        """
+        Adds a new %patch macro with given index and options.
+
+        If there are existing %patch macros, the new macro is added before,
+        after or between them according to index. Otherwise it is added
+        to the very end of %prep section.
+
+        Beware that it is valid to specify non-zero index and the -P option
+        at the same time, but the resulting macro behaves as two %patch macros
+        (even if both indices are the same, in such case the patch is applied
+        twice - you most likely don't want that).
+
+        Args:
+            index: Numeric index of the macro.
+            P: The -P option (patch index).
+            p: The -p option (strip number).
+            b: The -b option (backup).
+            E: The -E option (remove empty files).
+        """
+        macro = f"%patch{index}"
+        if P is not None:
+            macro += f" -P{P}"
+        if p is not None:
+            macro += f" -p{p}"
+        if b is not None:
+            macro += f" -b{b}"
+        if E:
+            macro += " -E"
+        macros = [m for m in self.macros if isinstance(m, PatchMacro)]
+        if macros:
+            lines = [
+                m._line
```

Comment:
Not a big fan of accessing the "private" attribute here. Perhaps `line` should be public?